lib/core/src/Cardano/Wallet/Api/Types.hs

```diff
@@ -1311,6 +1355,81 @@ instance (PassphraseMaxLength purpose, PassphraseMinLength purpose)
+instance ToJSON ApiScript where
+    toJSON (ApiScript (ApiT script')) =
+        object ["script" .= toJSON script']
+
+instance FromJSON ApiPubKey where
+    parseJSON = parseJSON >=> eitherToParser . first ShowFmt . fromText
+instance ToJSON ApiPubKey where
+    toJSON (ApiPubKey key') = do
+        let hrp = [Bech32.humanReadablePart|addr_vk|]
+        String $ T.decodeUtf8 $ encode (EBech32 hrp) key'
+
+instance FromJSON ApiCredential where
+    parseJSON obj = do
+        key' <-
+            (withObject "ApiCredential" $
+             \o -> o .:? "pub_key" :: Aeson.Parser (Maybe Text)) obj
+        case key' of
+            Nothing ->
+                CredentialScript <$> parseJSON obj
+            Just pubKeyTxt -> case fromText pubKeyTxt of
+                Right bytes ->
+                    pure $ CredentialPubKey bytes
+                Left (TextDecodingError err) -> fail err
+
+instance ToJSON ApiCredential where
+    toJSON (CredentialPubKey c)= object ["pub_key" .= toJSON c]
+    toJSON (CredentialScript c)= toJSON c
+
+instance FromJSON ApiAddressData where
+    parseJSON = withObject "ApiAddressData" $ \obj -> do
+        choice <- (,) <$> obj .:? "spending" <*> obj .:? "staking"
+        case choice of
+            (Nothing, Nothing) -> fail "ApiAddressData must have at least one credential."
+            (Just c, Nothing) -> do
+                spending' <- parseJSON c
+                pure $ AddrEnterprise spending'
+            (Nothing, Just c) -> do
+                staking' <- parseJSON c
+                pure $ AddrRewardAccount staking'
+            (Just c1, Just c2) -> do
+                spending' <- parseJSON c1
+                staking' <- parseJSON c2
+                pure $ AddrBase spending' staking'
+instance ToJSON ApiAddressData where
+    toJSON (AddrEnterprise spending') =
+        object [ "spending" .= toJSON spending']
+    toJSON (AddrRewardAccount staking') =
+        object [ "staking" .= toJSON staking']
+    toJSON (AddrBase spending' staking') =
+        object [ "spending" .= toJSON spending', "staking" .= toJSON staking']
+
+instance FromJSON AnyAddress where
+    parseJSON = withObject "AnyAddress" $ \obj -> do
+        addr <- obj .:? "address"
+        case addr of
+            Nothing ->
+                fail "AnyAddress expects 'address' key"
+            Just addr' -> case fromText addr' of
+                Right anyAddr -> pure anyAddr
+                Left (TextDecodingError err) -> fail err
+
+instance ToJSON AnyAddress where
+    toJSON (AnyAddress p addrType net) = do
+        let hrp = case (addrType, net) of
+                (EnterpriseDelegating, 1) -> [Bech32.humanReadablePart|addr|]
+                (RewardAccount, 1) -> [Bech32.humanReadablePart|stake|]
+                (EnterpriseDelegating, 0) -> [Bech32.humanReadablePart|addr_test|]
+                (RewardAccount, 0) -> [Bech32.humanReadablePart|stake_test|]
+                _ -> error "Wrong network tag in AnyAddress"
```

Comment:
It should be `fail` not `error`.

You might also be able to say that any network which isn't mainnet is a testnet.

Reply:
exposed network id in error. left `error` as I use it to get HRP. how would you use fail here?

Reply:
Ah OK I see -- this is the toJSON not the fromJSON.
This integer network tag thing seems fishy.
Normally we parse them into types at the program boundaries.
Anyway, I believe that it would be correct to use `_test` for any network that's not mainnet (or staging).
So I pushed a fix for that.